src/openssl.c

```diff
@@ -5487,6 +5488,85 @@ static int xr_setPublicKey(lua_State *L) {
 } /* xr_setPublicKey() */
 
 
+static int xr_setExtensionByNid(lua_State *L, X509_REQ *csr, int target_nid, void* value) {
+	STACK_OF(X509_EXTENSION) *sk = NULL;
+	X509_ATTRIBUTE *attr;
+	int has_attrs=0, idx, *pnid;
+
+	// Replace existing if it's there. Extensions are stored in a CSR in an interesting way:
+	//
+	// They are stored as a list under either (most likely) the "official"
+	// NID_ext_req or under NID_ms_ext_req which means everything is stored
+	// under a list in a single "attribute" so we can't use X509_REQ_add1_attr
+	// or similar.
+	//
+	// Instead we have to get the extensions, find and replace the SAN if it's
+	// in there, then *replace* the extensions in the list of attributes. (If
+	// we just try to add it the old ones are found first and don't take
+	// priority)
+
+	has_attrs = X509_REQ_get_attr_count(csr);
+	sk = X509_REQ_get_extensions(csr);
+
+	if (!X509V3_add1_i2d(&sk, target_nid, value, 0, X509V3_ADD_REPLACE))
+		goto error;
+
+	if (X509_REQ_add_extensions(csr, sk) == 0)
+		goto error;
+
+	// Delete the old extensions attribute, so that the one we just added takes priority
+	if (has_attrs) {
```

Comment:
Do you need to loop here in case there are several?
i.e. `for (; has_attrs>0; has_attrs--) {`

or does the loop below take care of that?
If so, why not just remove the `if (has_attrs) {` check (as `X509_REQ_get_extension_nids` should point to `NID_undef` on the first iteration)